optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
+                                SolverBenchmarkResult[] solverBenchmarkResultArray,
+                                long timeLeftPerSolverConfig) {
+        for (SolverBenchmarkResult solverBenchmarkResult : solverBenchmarkResultArray) {
             TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
             TerminationConfig tmpTerminationConfig = originalTerminationConfig == null
                     ? new TerminationConfig() : originalTerminationConfig.clone();
-            tmpTerminationConfig.shortenTimeMillisSpentLimit(timeLeft);
+            tmpTerminationConfig.overwriteTimeMillisSpentLimit(timeLeftPerSolverConfig);
             solverBenchmarkResult.getSolverConfig().setTerminationConfig(tmpTerminationConfig);
 
+            Integer singleBenchmarkResultIndex = singleBenchmarkResultIndexMap.get(solverBenchmarkResult);
+            if (singleBenchmarkResultIndex == null ||
+                    singleBenchmarkResultIndex >= solverBenchmarkResult.getSingleBenchmarkResultList().size()) {
+                singleBenchmarkResultIndex = 0;
+            }
+            SingleBenchmarkResult singleBenchmarkResult
+                    = solverBenchmarkResult.getSingleBenchmarkResultList().get(singleBenchmarkResultIndex);
+            if (singleBenchmarkResult != null) {
+                SingleBenchmarkRunner singleBenchmarkRunner = new SingleBenchmarkRunner(singleBenchmarkResult);
+                Future<SingleBenchmarkRunner> future = warmUpExecutorCompletionService.submit(singleBenchmarkRunner);
+                futureMap.put(future, singleBenchmarkRunner);
+                singleBenchmarkResultIndexMap.put(solverBenchmarkResult, singleBenchmarkResultIndex + 1);
+            }
+        }
+    }
+
+    private void warmUp(Map<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap,
+                        ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap, long timePhaseEnd) {
+        // Wait for the warm up benchmarks to complete
+        int tasksCount = futureMap.size();
+        for (int i = 0; i < tasksCount; i++) {
+            Future<SingleBenchmarkRunner> future;
             try {
-                Solver solver = solverBenchmarkResult.getSolverConfig().buildSolver();
-                solver.solve(problemBenchmarkResult.readPlanningProblem());
-            } catch (RuntimeException e) {
-                singleBenchmarkRunnerExceptionLogger.warn("The warmUp of singleBenchmark ("
-                        + singleBenchmarkResult.getName() + ") failed.", e);
-                // TODO update firstFailureSingleBenchmarkRunner (when warmups happen in a Runner too in parallel)
+                future = warmUpExecutorCompletionService.take();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
+                continue;
```

Comment:
A will - in most cases - effectively eat the exception (= sometimes goes wrong, but the developer/user doesn't see the notification). Eating the exception is about the worse thing than can happen, because it doesn't force the developer/user to fix the problem, it silently creates explainable problems which reduce long term thrust in our software. The benchmarker does uses the logger.error() itself once too, but that's only to delay the fail fast, instead of eating it: it still ends up throwing an exception instead of returning to the user. -1 on A (even with a delayed throw actually)

B is impossible/undesirable because InterruptedException is a checked exception and we don't want to make our benchmark users jumpt through some hoops of try-catch loops that they don't know what to do in the catch anyway. Big -1 on B

C in that form eats the exception, see reasoning of A. There might be a variation of this that is acceptable.

D should be a subclass of RuntimeException, such as IllegalStateException or IllegalArgumentException. I'd recommend IllegalStateException. +1 on D with an IllegalStateException and the InterruptedException on the cause

E will cause an infinite loop. Big -1 on E